server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -945,13 +974,16 @@ public void prepare() {
 
             if (ignoredValues != null && ignoredValues.isEmpty() == false) {
                 for (IgnoredSourceFieldMapper.NameValue value : ignoredValues) {
+                    if (filter != null && filter.isPathFiltered(value.name(), XContentDataHelper.isEncodedObject(value.value()))) {
```

Comment:
These extra calls to `isEncodedObject` can be expensive, iiuc.. If so, we should decode once and cache the result, or use different encodings in `XContentDataHelper` for arrays vs objects.

Reply:
Also, should this filtering be happening earlier, in https://github.com/elastic/elasticsearch/blob/b9855b8e4e0f29069a50f4e41a4643beb532c75f/server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java#L201

Reply:
I agree, i think it is cleaner to do this in `decode`.

Reply:
I simplified the detection in https://github.com/elastic/elasticsearch/pull/113827/commits/886506067380d370163848a9c681665480a1e2e9 . We only need to know if the value might contain sub-object. It's ok if we're wrong (e.g.: an array of primitives) since the xcontent filtering will remove the value for us. So this is more an optimisation to avoid the parsing now.